include/deal.II/lac/trilinos_tpetra_vector.h

```diff
@@ -0,0 +1,406 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2015 - 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_trilinos_tpetra_vector_h
+#define dealii_trilinos_tpetra_vector_h
+
+
+#include <deal.II/base/config.h>
+
+#if defined(DEAL_II_WITH_TRILINOS) && defined(DEAL_II_WITH_MPI)
+
+#  include <deal.II/base/index_set.h>
+#  include <deal.II/base/subscriptor.h>
+
+#  include <deal.II/lac/trilinos_tpetra_communication_pattern.h>
+#  include <deal.II/lac/vector_operation.h>
+#  include <deal.II/lac/vector_space_vector.h>
+#  include <deal.II/lac/vector_type_traits.h>
+
+#  include <Teuchos_Comm.hpp>
+#  include <Teuchos_OrdinalTraits.hpp>
+#  include <Tpetra_Core.hpp>
+#  include <Tpetra_Vector.hpp>
+#  include <Tpetra_Version.hpp>
+#  include <mpi.h>
+
+#  include <memory>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace LinearAlgebra
+{
+  // Forward declaration
+  template <typename Number>
+  class ReadWriteVector;
+
+  namespace TpetraWrappers
+  {
+    /**
+     * This class implements a wrapper to the Trilinos distributed vector
+     * class Tpetra::Vector. This class is derived from the
+     * LinearAlgebra::VectorSpaceVector class. Note however that Tpetra only
+     * works with Number = double.
+     *
+     * @ingroup TrilinosWrappers
+     * @ingroup Vectors
+     * @author Daniel Arndt, 2018
+     */
+    class Vector : public VectorSpaceVector<double>, public Subscriptor
```

Comment:
my main comment so far is I think we should start from defining `VectorSpaceMultiVector` as this is what naturally supported by TPetra. Note that this is NOT just a `std::vector<VectorSpaceVector>` as it's the multivector which is responsible for all memory allocation etc, and it may (or may not) provide a way to get one of those vectors.

Think of problems with multiple RHS or problems with subspaces.

I think it's crucial to start form that end as otherwise the `Vector` here is still reponsible for managing its memory (it's essentially a multivector with a single column), but if you start from higher level, then `VectorSpaceMultiVector` may have methods which return a _view_ to a column(s) or return _copy_ of a colum(s).

Reply:
that's how you get a vector out of multivector: 

https://trilinos.org/docs/dev/packages/tpetra/doc/html/classTpetra_1_1MultiVector.html#a7951dcc07f3da968cb2b9b9a4d5fdcdc